internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,169 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/openshift/assisted-service/internal/bminventory"
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// InstallEnvReconciler reconciles a InstallEnv object
+type InstallEnvReconciler struct {
+	client.Client
+	Log                      logrus.FieldLogger
+	Scheme                   *runtime.Scheme
+	Installer                bminventory.InstallerInternals
+	PullSecretUpdatesChannel chan event.GenericEvent
+}
+
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs/status,verbs=get;update;patch
+
+func (r *InstallEnvReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	installEnv := &adiiov1alpha1.InstallEnv{}
+	if err := r.Get(ctx, req.NamespacedName, installEnv); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.create(ctx, installEnv)
+}
+
+func (r *InstallEnvReconciler) create(ctx context.Context, installEnv *adiiov1alpha1.InstallEnv) (ctrl.Result, error) {
+	var inventoryErr error
+	var imageInfo *models.ImageInfo
+	var updatedCluster *common.Cluster
+	state := adiiov1alpha1.ImageStateFailedToCreate
+
+	clusterDeployment, clusterDeploymentRefErr := r.getClusterDeploymentByRef(ctx, installEnv.Spec.ClusterRef)
+	if clusterDeploymentRefErr != nil {
+		return r.updateStatusAndReturnResult(ctx, installEnv, nil, state, clusterDeploymentRefErr)
+	}
+
+	c, err := r.Installer.GetClusterByKubeKey(types.NamespacedName{
+		Name:      clusterDeployment.Name,
+		Namespace: clusterDeployment.Namespace,
+	})
+
+	if gorm.IsRecordNotFoundError(err) {
+		inventoryErr = common.NewApiError(http.StatusNotFound, err)
+
+	} else {
+		isoParams := installer.GenerateClusterISOParams{
+			ClusterID: *c.ID,
+			ImageCreateParams: &models.ImageCreateParams{
+			},
+		}
+		if len(installEnv.Spec.SSHAuthorizedKeys) > 0 {
+			isoParams.ImageCreateParams.SSHPublicKey = installEnv.Spec.SSHAuthorizedKeys[0]
+		}
+		updatedCluster, inventoryErr = r.Installer.GenerateClusterISOInternal(ctx, isoParams)
+	}
+
+	if inventoryErr == nil {
+		imageInfo = updatedCluster.ImageInfo
+		state = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, installEnv, imageInfo, state, inventoryErr)
+}
+
+func (r *InstallEnvReconciler) getClusterDeploymentByRef(ctx context.Context, ref *adiiov1alpha1.ClusterReference) (*hivev1.ClusterDeployment, error) {
```

Comment:
This method is the same as the controller-runtime client's `Get` method with two main differences: its reference struct has a different name, and it wraps the error. The "by reference" part of this method is not that interesting; it's really the error wrapping that constitutes the unique behavior. So at a minimum I'd consider a rename of the method.

But rather than make a whole new "get" method, consider just using the normal client's Get method directly in-line where you need it, and having a separate method just to help wrap the error. I think devs who have worked on other controller code would find that more familiar and readable.